contracts/lease/src/position/mod.rs

```diff
@@ -1,2 +1,111 @@
+use currency::Currency;
 pub use dto::PositionDTO;
+use finance::{
+    coin::Coin,
+    liability::{Liability, Status},
+    price::{self, Price},
+};
+
+use crate::{
+    api::{LeaseCoin, PositionSpec},
+    error::{ContractError, ContractResult},
+};
+
 mod dto;
+
+#[cfg_attr(test, derive(Debug))]
+pub struct Position<Asset, Lpn> {
+    amount: Coin<Asset>,
+    liability: Liability,
+    min_asset: Coin<Lpn>,
+    min_sell_asset: Coin<Lpn>,
+}
+
+impl<Asset, Lpn> Position<Asset, Lpn>
+where
+    Asset: Currency,
+    Lpn: Currency,
+{
+    fn new_internal(
+        amount: Coin<Asset>,
+        liability: Liability,
+        min_asset: Coin<Lpn>,
+        min_sell_asset: Coin<Lpn>,
+    ) -> Self {
+        let obj = Self {
+            amount,
+            liability,
+            min_asset,
+            min_sell_asset,
+        };
+        debug_assert_eq!(Ok(()), obj.invariant_held());
+        obj
+    }
+
+    pub fn try_from(amount: &LeaseCoin, spec: PositionSpec) -> ContractResult<Self> {
```

Comment:
it's more natural in from/into functionality to pass the data by value
please change the amount type and make the necessary changes for the amount value to get to this function